dev/buildtool/inspection_commands.py

```diff
@@ -12,100 +12,1154 @@
-        source_repositories)
-    if self.options.halyard_version:
-      repositories = list(repositories)
-      repositories.append(GitRepositorySpec('halyard'))
-
-    return repositories
-
-  def _do_repository(self, repository):
-    decorator = 'spinnaker-' if repository.name != 'spinnaker' else ''
-    if repository.name == 'halyard':
-      name = 'halyard'
-      name_version = 'halyard_' + self.options.halyard_version
+    if options.bintray_org is None != options.bintray_debian_repository is None:
+      raise_and_log_error(
+          ConfigError('Either neither or both "bintray_org"'
+                      ' and "bintray_debian_repository" should be specified'))
+
+    self.__bad_files = {}
+    self.__non_standard_boms = []
+
+    # We're going to have a bunch of threads each writing into different keys
+    # in order to deconflict with one another lockless. Then we'll aggregate
+    # it all together when we're done processing for a single aggregate result.
+    self.__per_thread_result_map = {}
+
+    self.__expect_docker_registry = options.docker_registry
+    self.__expect_debian_repository = (
+        'https://dl.bintray.com/%s/%s' % (options.bintray_org,
+                                          options.bintray_debian_repository)
+        if options.bintray_org
+        else None)
+
+    super(CollectBomVersions, self).__init__(
+        factory, options, **kwargs)
+
+  def load_bom_from_url(self, url):
+    """Returns the bom specification dict from a gcs url."""
+    logging.debug('Loading %s', url)
+    try:
+      text = check_subprocess('gsutil cat ' + url)
+      return yaml.load(text)
+    except Exception as ex:
+      self.__bad_files[url] = ex.message
+      maybe_log_exception('load_from_from_url', ex,
+                          action_msg='Skipping %s' % url)
+      return None
+
+  def extract_bom_info(self, bom):
+    """Return a minimal dict identifying this BOM.
+
+    This also includes non-standard config specified by this BOM.
+    """
+    info = {
+        'bom_version': bom['version'],
+        'bom_timestamp': bom.get('timestamp', 'NotRecorded')
+    }
+    artifact_sources = bom.get('artifactSources')
+    if artifact_sources is None:
+      logging.warning('%s does not have artifactSources', bom['version'])
+      return info
+
+    def add_if_nonstandard(name, expect):
+      if artifact_sources[name] != expect:
+        logging.warning('%s has nonstandard %s = %s',
+                        bom['version'], name, artifact_sources[name])
+        info[name] = artifact_sources[name]
+
+    add_if_nonstandard('dockerRegistry', self.__expect_docker_registry)
+    add_if_nonstandard('debianRepository', self.__expect_debian_repository)
+    if len(info) > 2:
+      self.__non_standard_boms.append(info)
+
+    return info
+
+  def analyze_bom(self, bom):
+    """Analyzes one bom and breaks it down into this threads result_map.
+
+    Boms are processed within a single thread, but multiple boms can be
```

Comment:
I do check the file name matches the version and report that among the "invalid boms". This does not seem to happen in practice because halyard writes it. It could happen if you deposit the boms directly in the bucket. But again, it would get picked up elsewhere.